# introducao_br — Transfer Learning (PT-BR)

Este notebook demonstra **Transfer Learning / Fine-Tuning** com Keras/TensorFlow usando um **modelo pré-treinado** (ex.: `VGG16`) como extrator de características e, opcionalmente, ajustando as últimas camadas (*fine-tuning*).

## O que você vai ver aqui
- **Carregamento de dados** (imagens e rótulos) e pré-processamento no formato esperado pelo modelo pré-treinado.
- **Feature extraction**: congelamos o backbone (ex.: `VGG16`) e treinamos **apenas a cabeça** (camada(s) final(is)).
- **Fine-tuning (opcional)**: descongelamos **poucas camadas finais** e treinamos com **taxa de aprendizado menor**.
- **Métricas**: `accuracy`, `val_accuracy`, `loss`, `val_loss` a cada época.

## Como interpretar os logs
- `accuracy`/`loss`: desempenho no **treino**.
- `val_accuracy`/`val_loss`: desempenho na **validação** (o mais importante para generalização).
- Se `loss` ↓ e `val_loss` ↓ → ótimo. Se `loss` ↓ e `val_loss` ↑ de forma consistente → **overfitting**.

## Dicas práticas
- Comece com **feature extraction** (backbone congelado).
- No *fine-tuning*, **libere poucas camadas finais** e **reduza** a *learning rate*.
- Em **CPU**, prefira `batch_size` menor (8–32) e imagens 160–224 px.

> **Observação**: todo o conteúdo **original** do notebook foi **preservado**.  
> Abaixo foram adicionadas **apenas duas alterações** para melhorar a execução em **CPU**, marcadas com **`[ALTERAÇÃO CPU]`**.


In [ ]:
# [ALTERAÇÃO CPU] Forçar uso de CPU e backend do Keras
import os
os.environ["KERAS_BACKEND"] = "tensorflow"  # Keras 3 -> usar TensorFlow como backend

import tensorflow as tf
try:
    # Oculta GPUs (inclui DirectML, se houver) para rodar de forma estável na CPU
    tf.config.set_visible_devices([], 'GPU')
    print("CPU forçada: GPUs ocultas com sucesso.")
except Exception as e:
    print("Aviso: não foi possível ocultar GPU (ok se não houver GPU):", e)

print("TF:", tf.__version__)
print("Dispositivos visíveis:", tf.config.list_physical_devices())


In [ ]:
# [ALTERAÇÃO CPU] Otimização leve de I/O: aplicar prefetch se os datasets já existirem
try:
    AUTOTUNE = tf.data.AUTOTUNE
    for name in ["train_ds", "train_dataset", "ds_train"]:
        if name in globals():
            globals()[name] = globals()[name].prefetch(AUTOTUNE)
    for name in ["val_ds", "validation_ds", "ds_val", "test_ds", "ds_test"]:
        if name in globals():
            globals()[name] = globals()[name].prefetch(AUTOTUNE)
    print("Prefetch ativado quando datasets foram encontrados.")
except Exception as e:
    print("Prefetch não aplicado (ok):", e)


# Transfer learning / fine-tuning

This tutorial will guide you through the process of using _transfer learning_ to learn an accurate image classifier from a relatively small number of training samples. Generally speaking, transfer learning refers to the process of leveraging the knowledge learned in one model for the training of another model.

More specifically, the process involves taking an existing neural network which was previously trained to good performance on a larger dataset, and using it as the basis for a new model which leverages that previous network's accuracy for a new task. This method has become popular in recent years to improve the performance of a neural net trained on a small dataset; the intuition is that the new dataset may be too small to train to good performance by itself, but we know that most neural nets trained to learn image features often learn similar features anyway, especially at early layers where they are more generic (edge detectors, blobs, and so on).

Transfer learning has been largely enabled by the open-sourcing of state-of-the-art models; for the top performing models in image classification tasks (like from [ILSVRC](http://www.image-net.org/challenges/LSVRC/)), it is common practice now to not only publish the architecture, but to release the trained weights of the model as well. This lets amateurs use these top image classifiers to boost the performance of their own task-specific models.

#### Feature extraction vs. fine-tuning

At one extreme, transfer learning can involve taking the pre-trained network and freezing the weights, and using one of its hidden layers (usually the last one) as a feature extractor, using those features as the input to a smaller neural net.

At the other extreme, we start with the pre-trained network, but we allow some of the weights (usually the last layer or last few layers) to be modified. Another name for this procedure is called "fine-tuning" because we are slightly adjusting the pre-trained net's weights to the new task. We usually train such a network with a lower learning rate, since we expect the features are already relatively good and do not need to be changed too much.

Sometimes, we do something in-between: Freeze just the early/generic layers, but fine-tune the later layers. Which strategy is best depends on the size of your dataset, the number of classes, and how much it resembles the dataset the previous model was trained on (and thus, whether it can benefit from the same learned feature extractors). A more detailed discussion of how to strategize can be found in [[1]](http://cs231n.github.io/transfer-learning/) [[2]](http://sebastianruder.com/transfer-learning/).

## Procedure

In this guide will go through the process of loading a state-of-the-art, 1000-class image classifier, [VGG16](https://arxiv.org/pdf/1409.1556.pdf) which [won the ImageNet challenge in 2014](http://www.robots.ox.ac.uk/~vgg/research/very_deep/), and using it as a fixed feature extractor to train a smaller custom classifier on our own images, although with very few code changes, you can try fine-tuning as well.

We will first load VGG16 and remove its final layer, the 1000-class softmax classification layer specific to ImageNet, and replace it with a new classification layer for the classes we are training over. We will then freeze all the weights in the network except the new ones connecting to the new classification layer, and then train the new classification layer over our new dataset.

We will also compare this method to training a small neural network from scratch on the new dataset, and as we shall see, it will dramatically improve our accuracy. We will do that part first.

As our test subject, we'll use a dataset consisting of around 6000 images belonging to 97 classes, and train an image classifier with around 80% accuracy on it. It's worth noting that this strategy scales well to image sets where you may have even just a couple hundred or less images. Its performance will be lesser from a small number of samples (depending on classes) as usual, but still impressive considering the usual constraints.


In [1]:
import sys, numpy
print(sys.executable, numpy.__version__)


C:\Users\poger\.conda\envs\AI\python.exe 1.26.4


In [ ]:
#from google.colab import drive
#drive.mount('/content/drive')

In [2]:
import sys, platform, os
print("Python exe:", sys.executable)
print("sys.version:", sys.version)
print("sys.prefix:", sys.prefix)
print("Platform:", platform.platform())


Python exe: C:\Users\poger\.conda\envs\AI\python.exe
sys.version: 3.12.3 | packaged by conda-forge | (main, Apr 15 2024, 18:20:11) [MSC v.1938 64 bit (AMD64)]
sys.prefix: C:\Users\poger\.conda\envs\AI
Platform: Windows-10-10.0.19045-SP0


In [3]:
# o que já está instalado neste kernel?
%pip show keras tensorflow



Name: keras
Version: 3.11.2
Summary: Multi-backend Keras
Home-page: 
Author: 
Author-email: Keras team <keras-users@googlegroups.com>
License: Apache License 2.0
Location: C:\Users\poger\.conda\envs\AI\Lib\site-packages
Requires: absl-py, h5py, ml-dtypes, namex, numpy, optree, packaging, rich
Required-by: tensorflow, tensorflow_cpu
---
Name: tensorflow
Version: 2.20.0
Summary: TensorFlow is an open source machine learning framework for everyone.
Home-page: https://www.tensorflow.org/
Author: Google Inc.
Author-email: packages@tensorflow.org
License: Apache 2.0
Location: C:\Users\poger\.conda\envs\AI\Lib\site-packages
Requires: absl-py, astunparse, flatbuffers, gast, google_pasta, grpcio, h5py, keras, libclang, ml_dtypes, numpy, opt_einsum, packaging, protobuf, requests, setuptools, six, tensorboard, termcolor, typing_extensions, wrapt
Required-by: 
Note: you may need to restart the kernel to use updated packages.


In [5]:
%matplotlib inline

import os

#if using Theano with GPU
#os.environ["KERAS_BACKEND"] = "tensorflow"

import random
import numpy as np
import keras

import matplotlib.pyplot as plt
from matplotlib.pyplot import imshow

from keras.preprocessing import image
from keras.applications.imagenet_utils import preprocess_input
from keras.models import Sequential
from keras.layers import Dense, Dropout, Flatten, Activation
from keras.layers import Conv2D, MaxPooling2D
from keras.models import Model

### Getting a dataset

The first step is going to be to load our data. As our example, we will be using the dataset [CalTech-101](http://www.vision.caltech.edu/Image_Datasets/Caltech101/), which contains around 9000 labeled images belonging to 101 object categories. However, we will exclude 5 of the categories which have the most images. This is in order to keep the class distribution fairly balanced (around 50-100) and constrained to a smaller number of images, around 6000.

To obtain this dataset, you can either run the download script `download.sh` in the `data` folder, or the following commands:

    wget http://www.vision.caltech.edu/Image_Datasets/Caltech101/101_ObjectCategories.tar.gz
    tar -xvzf 101_ObjectCategories.tar.gz

If you wish to use your own dataset, it should be aranged in the same fashion to `101_ObjectCategories` with all of the images organized into subfolders, one for each class. In this case, the following cell should load your custom dataset correctly by just replacing `root` with your folder. If you have an alternate structure, you just need to make sure that you load the list `data` where every element is a dict where `x` is the data (a 1-d numpy array) and `y` is the label (an integer). Use the helper function `get_image(path)` to load the image correctly into the array, and note also that the images are being resized to 224x224. This is necessary because the input to VGG16 is a 224x224 RGB image. You do not need to resize them on your hard drive, as that is being done in the code below.

If you have `101_ObjectCategories` in your data folder, the following cell should load all the data.

In [6]:
!pip -q install gdown

# ID do arquivo no seu link: 1bCJaQE-rVER2JwbaYwwjzNO_AuXDikbQ
!echo "Downloading 101_Object_Categories for image notebooks"
!gdown --id 1bCJaQE-rVER2JwbaYwwjzNO_AuXDikbQ -O 101_ObjectCategories.tar.gz
!tar -xzf 101_ObjectCategories.tar.gz
!rm 101_ObjectCategories.tar.gz
!ls



[notice] A new release of pip is available: 24.3.1 -> 25.2
[notice] To update, run: python.exe -m pip install --upgrade pip


"Downloading 101_Object_Categories for image notebooks"


C:\Users\poger\.conda\envs\AI\Lib\site-packages\gdown\__main__.py:140: FutureWarning: Option `--id` was deprecated in version 4.3.1 and will be removed in 5.0. You don't need to pass it anymore to use a file ID.
  warnings.warn(
Downloading...
From (original): https://drive.google.com/uc?id=1bCJaQE-rVER2JwbaYwwjzNO_AuXDikbQ
From (redirected): https://drive.google.com/uc?id=1bCJaQE-rVER2JwbaYwwjzNO_AuXDikbQ&confirm=t&uuid=b3e1e03a-523b-49dd-b8f4-e1caa723bedc
To: C:\Users\poger\101_ObjectCategories.tar.gz

  0%|          | 0.00/132M [00:00<?, ?B/s]
  0%|          | 524k/132M [00:01<05:31, 396kB/s]
  1%|          | 1.05M/132M [00:02<05:47, 376kB/s]
  1%|1         | 1.57M/132M [00:03<05:10, 419kB/s]
  2%|1         | 2.10M/132M [00:04<04:31, 477kB/s]
  2%|1         | 2.62M/132M [00:05<04:24, 488kB/s]
  2%|2         | 3.15M/132M [00:07<05:18, 404kB/s]
  3%|2         | 3.67M/132M [00:07<04:09, 514kB/s]
  3%|3         | 4.19M/132M [00:10<06:41, 317kB/s]
  4%|3         | 4.72M/132M [00:14<09:18

-1.14-windows.xml
101_ObjectCategories
1_lab1.ipynb
3D Objects
Ambiente de Impressão
Ambiente de Rede
AppData
Application Data
CascadeProjects
Configurações Locais
Contacts
Cookies
Dados de Aplicativos
Desktop
Documents
Downloads
Exercicios
Favorites
InstallAnywhere
IntelGraphicsProfiles
Links
Menu Iniciar
Meus Documentos
Modelos
Music
NTUSER.DAT
NTUSER.DAT{1a566ac0-8f5d-11f0-8a47-005056c00008}.TM.blf
NTUSER.DAT{1a566ac0-8f5d-11f0-8a47-005056c00008}.TMContainer00000000000000000001.regtrans-ms
NTUSER.DAT{1a566ac0-8f5d-11f0-8a47-005056c00008}.TMContainer00000000000000000002.regtrans-ms
OneDrive
Oracle
Pictures
Postman
Postman Agent
Recent
Saved Games
Searches
SendTo
Sti_Trace.log
Untitled.ipynb
Videos
ai_completion
anaconda_projects
ansel
cachedir_joblib
chroma.log
delete
demo_find_face.py
demo_find_green_blob.py
dwhelper
helloworld_1.py
ipython_hist.sqlite
json
maxima
migrations_202504181229.txt
my_chroma_data
ntuser.dat.LOG1
ntuser.dat.LOG2
ntuser.ini
requirements.txt
scoop
source
s

In [7]:
import os
print("CWD:", os.getcwd())   # diretório atual


CWD: C:\Users\poger


In [8]:
root = '101_ObjectCategories'
exclude = ['BACKGROUND_Google', 'Motorbikes', 'airplanes', 'Faces_easy', 'Faces']
train_split, val_split = 0.7, 0.15

categories = [x[0] for x in os.walk(root) if x[0]][1:]
categories = [c for c in categories if c not in [os.path.join(root, e) for e in exclude]]

print(categories)
print(f"Contents of {root}: {os.listdir(root)}")

['101_ObjectCategories\\accordion', '101_ObjectCategories\\anchor', '101_ObjectCategories\\ant', '101_ObjectCategories\\barrel', '101_ObjectCategories\\bass', '101_ObjectCategories\\beaver', '101_ObjectCategories\\binocular', '101_ObjectCategories\\bonsai', '101_ObjectCategories\\brain', '101_ObjectCategories\\brontosaurus', '101_ObjectCategories\\buddha', '101_ObjectCategories\\butterfly', '101_ObjectCategories\\camera', '101_ObjectCategories\\cannon', '101_ObjectCategories\\car_side', '101_ObjectCategories\\ceiling_fan', '101_ObjectCategories\\cellphone', '101_ObjectCategories\\chair', '101_ObjectCategories\\chandelier', '101_ObjectCategories\\cougar_body', '101_ObjectCategories\\cougar_face', '101_ObjectCategories\\crab', '101_ObjectCategories\\crayfish', '101_ObjectCategories\\crocodile', '101_ObjectCategories\\crocodile_head', '101_ObjectCategories\\cup', '101_ObjectCategories\\dalmatian', '101_ObjectCategories\\dollar_bill', '101_ObjectCategories\\dolphin', '101_ObjectCategories\

This function is useful for pre-processing the data into an image and input vector.

In [9]:
# helper function to load image and return it and input vector
def get_image(path):
    img = image.load_img(path, target_size=(224, 224))
    x = image.img_to_array(img)
    x = np.expand_dims(x, axis=0)
    x = preprocess_input(x)
    return img, x

Load all the images from root folder

In [10]:
data = []
for c, category in enumerate(categories):
    images = [os.path.join(dp, f) for dp, dn, filenames
              in os.walk(category) for f in filenames
              if os.path.splitext(f)[1].lower() in ['.jpg','.png','.jpeg']]
    for img_path in images:
        img, x = get_image(img_path)
        data.append({'x':np.array(x[0]), 'y':c})

# count the number of classes
num_classes = len(categories)

Randomize the data order.

In [11]:
random.shuffle(data)

create training / validation / test split (70%, 15%, 15%)

In [12]:
idx_val = int(train_split * len(data))
idx_test = int((train_split + val_split) * len(data))
train = data[:idx_val]
val = data[idx_val:idx_test]
test = data[idx_test:]

Separate data for labels.

In [13]:
x_train, y_train = np.array([t["x"] for t in train]), [t["y"] for t in train]
x_val, y_val = np.array([t["x"] for t in val]), [t["y"] for t in val]
x_test, y_test = np.array([t["x"] for t in test]), [t["y"] for t in test]
print(y_test)

[53, 96, 38, 54, 22, 6, 46, 89, 72, 56, 48, 68, 33, 27, 41, 57, 95, 72, 86, 19, 10, 96, 6, 79, 4, 53, 44, 56, 30, 61, 17, 53, 78, 53, 90, 52, 19, 26, 79, 76, 56, 68, 44, 75, 54, 39, 34, 14, 4, 55, 80, 27, 35, 62, 87, 91, 62, 73, 86, 94, 57, 20, 70, 67, 28, 82, 52, 51, 22, 45, 38, 5, 45, 53, 8, 50, 70, 90, 66, 90, 86, 52, 27, 94, 7, 53, 49, 40, 11, 35, 88, 11, 5, 50, 52, 59, 6, 90, 64, 47, 7, 7, 22, 26, 53, 26, 73, 61, 34, 4, 79, 73, 81, 38, 29, 14, 45, 21, 47, 18, 82, 86, 68, 84, 65, 0, 83, 31, 64, 11, 22, 78, 85, 90, 85, 5, 30, 11, 9, 33, 93, 88, 22, 72, 70, 20, 42, 90, 90, 70, 26, 66, 23, 65, 48, 68, 23, 3, 82, 44, 14, 61, 14, 25, 14, 46, 27, 73, 92, 41, 59, 88, 68, 3, 22, 37, 43, 64, 19, 52, 11, 90, 17, 15, 71, 11, 53, 96, 78, 3, 44, 90, 46, 72, 65, 64, 37, 43, 90, 8, 84, 71, 31, 77, 52, 62, 11, 86, 53, 54, 11, 16, 4, 46, 25, 7, 8, 18, 12, 3, 69, 42, 17, 71, 43, 54, 16, 32, 78, 73, 49, 12, 52, 7, 53, 23, 89, 9, 10, 52, 50, 11, 26, 65, 22, 10, 24, 90, 63, 77, 51, 37, 36, 32, 85, 3, 9

Pre-process the data as before by making sure it's float32 and normalized between 0 and 1.

In [14]:
# normalize data
x_train = x_train.astype('float32') / 255.
x_val = x_val.astype('float32') / 255.
x_test = x_test.astype('float32') / 255.

# convert labels to one-hot vectors
y_train = keras.utils.to_categorical(y_train, num_classes)
y_val = keras.utils.to_categorical(y_val, num_classes)
y_test = keras.utils.to_categorical(y_test, num_classes)
print(y_test.shape)

(932, 97)


Let's get a summary of what we have.

In [15]:
# summary
print("finished loading %d images from %d categories"%(len(data), num_classes))
print("train / validation / test split: %d, %d, %d"%(len(x_train), len(x_val), len(x_test)))
print("training data shape: ", x_train.shape)
print("training labels shape: ", y_train.shape)


finished loading 6209 images from 97 categories
train / validation / test split: 4346, 931, 932
training data shape:  (4346, 224, 224, 3)
training labels shape:  (4346, 97)


If everything worked properly, you should have loaded a bunch of images, and split them into three sets: `train`, `val`, and `test`. The shape of the training data should be (`n`, 224, 224, 3) where `n` is the size of your training set, and the labels should be (`n`, `c`) where `c` is the number of classes (97 in the case of `101_ObjectCategories`.

Notice that we divided all the data into three subsets -- a training set `train`, a validation set `val`, and a test set `test`. The reason for this is to properly evaluate the accuracy of our classifier. During training, the optimizer uses the validation set to evaluate its internal performance, in order to determine the gradient without overfitting to the training set. The `test` set is always held out from the training algorithm, and is only used at the end to evaluate the final accuracy of our model.

Let's quickly look at a few sample images from our dataset.

In [16]:
images = [os.path.join(dp, f) for dp, dn, filenames in os.walk(root) for f in filenames if os.path.splitext(f)[1].lower() in ['.jpg','.png','.jpeg']]
idx = [int(len(images) * random.random()) for i in range(8)]
imgs = [image.load_img(images[i], target_size=(224, 224)) for i in idx]
concat_image = np.concatenate([np.asarray(img) for img in imgs], axis=1)
plt.figure(figsize=(16,4))
plt.imshow(concat_image)

### First training a neural net from scratch

Before doing the transfer learning, let's first build a neural network from scratch for doing classification on our dataset. This will give us a baseline to compare to our transfer-learned network later.

The network we will construct contains 4 alternating convolutional and max-pooling layers, followed by a [dropout](https://www.cs.toronto.edu/~hinton/absps/JMLRdropout.pdf) after every other conv/pooling pair. After the last pooling layer, we will attach a fully-connected layer with 256 neurons, another dropout layer, then finally a softmax classification layer for our classes.

Our loss function will be, as usual, categorical cross-entropy loss, and our learning algorithm will be [AdaDelta](https://arxiv.org/abs/1212.5701). Various things about this network can be changed to get better performance, perhaps using a larger network or a different optimizer will help, but for the purposes of this notebook, the goal is to just get an understanding of an approximate baseline for comparison's sake, and so it isn't neccessary to spend much time trying to optimize this network.

Upon compiling the network, let's run `model.summary()` to get a snapshot of its layers.

In [17]:
# build the network
model = Sequential()
print("Input dimensions: ",x_train.shape[1:])

model.add(Conv2D(32, (3, 3), input_shape=x_train.shape[1:]))
model.add(Activation('relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))

model.add(Conv2D(32, (3, 3)))
model.add(Activation('relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))

model.add(Dropout(0.25))

model.add(Conv2D(32, (3, 3)))
model.add(Activation('relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))

model.add(Conv2D(32, (3, 3)))
model.add(Activation('relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))

model.add(Dropout(0.25))

model.add(Flatten())
model.add(Dense(256))
model.add(Activation('relu'))

model.add(Dropout(0.5))

model.add(Dense(num_classes))
model.add(Activation('softmax'))

model.summary()

Input dimensions:  (224, 224, 3)


C:\Users\poger\.conda\envs\AI\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:113: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ conv2d (Conv2D)                      │ (None, 222, 222, 32)        │             896 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ activation (Activation)              │ (None, 222, 222, 32)        │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ max_pooling2d (MaxPooling2D)         │ (None, 111, 111, 32)        │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ conv2d_1 (Conv2D)                    │ (None, 109, 109, 32)        │           9,248 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ activation_1 (Activation)            │ (None, 109, 109, 32)        │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ max_pooling2d_1 (MaxPooling2D)       │ (None, 54, 54, 32)          │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dropout (Dropout)                    │ (None, 54, 54, 32)          │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ conv2d_2 (Conv2D)                    │ (None, 52, 52, 32)          │           9,248 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ activation_2 (Activation)            │ (None, 52, 52, 32)          │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ max_pooling2d_2 (MaxPooling2D)       │ (None, 26, 26, 32)          │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ conv2d_3 (Conv2D)                    │ (None, 24, 24, 32)          │           9,248 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ activation_3 (Activation)            │ (None, 24, 24, 32)          │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ max_pooling2d_3 (MaxPooling2D)       │ (None, 12, 12, 32)          │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dropout_1 (Dropout)                  │ (None, 12, 12, 32)          │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ flatten (Flatten)                    │ (None, 4608)                │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense (Dense)                        │ (None, 256)                 │       1,179,904 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ activation_4 (Activation)            │ (None, 256)                 │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dropout_2 (Dropout)                  │ (None, 256)                 │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_1 (Dense)                      │ (None, 97)                  │          24,929 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ activation_5 (Activation)            │ (None, 97)                  │               0 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 1,233,473 (4.71 MB)

 Trainable params: 1,233,473 (4.71 MB)

 Non-trainable params: 0 (0.00 B)

We've created a medium-sized network with ~1.2 million weights and biases (the parameters). Most of them are leading into the one pre-softmax fully-connected layer "dense_5".

We can now go ahead and train our model for 100 epochs with a batch size of 128. We'll also record its history so we can plot the loss over time later.

In [18]:
# compile the model to use categorical cross-entropy loss function and adadelta optimizer
model.compile(loss='categorical_crossentropy',
              optimizer='adam',
              metrics=['accuracy'])

history = model.fit(x_train, y_train,
                    batch_size=128,
                    epochs=10,
                    validation_data=(x_val, y_val))


Epoch 1/10
34/34 ━━━━━━━━━━━━━━━━━━━━ 113s 3s/step - accuracy: 0.0423 - loss: 4.5027 - val_accuracy: 0.0763 - val_loss: 4.3943
Epoch 2/10
34/34 ━━━━━━━━━━━━━━━━━━━━ 85s 3s/step - accuracy: 0.0927 - loss: 4.2709 - val_accuracy: 0.1343 - val_loss: 4.1373
Epoch 3/10
34/34 ━━━━━━━━━━━━━━━━━━━━ 82s 2s/step - accuracy: 0.1608 - loss: 3.9326 - val_accuracy: 0.2234 - val_loss: 3.7618
Epoch 4/10
34/34 ━━━━━━━━━━━━━━━━━━━━ 81s 2s/step - accuracy: 0.2218 - loss: 3.5381 - val_accuracy: 0.2707 - val_loss: 3.4405
Epoch 5/10
34/34 ━━━━━━━━━━━━━━━━━━━━ 78s 2s/step - accuracy: 0.2862 - loss: 3.1493 - val_accuracy: 0.3050 - val_loss: 3.1216
Epoch 6/10
34/34 ━━━━━━━━━━━━━━━━━━━━ 83s 2s/step - accuracy: 0.3426 - loss: 2.8003 - val_accuracy: 0.3673 - val_loss: 2.8687
Epoch 7/10
34/34 ━━━━━━━━━━━━━━━━━━━━ 77s 2s/step - accuracy: 0.4082 - loss: 2.4916 - val_accuracy: 0.3792 - val_loss: 2.7824
Epoch 8/10
34/34 ━━━━━━━━━━━━━━━━━━━━ 80s 2s/step - accuracy: 0.4494 - loss: 2.2379 - val_accuracy: 0.4006 - val_loss

Let's plot the validation loss and validation accuracy over time.

In [19]:
fig = plt.figure(figsize=(16,4))
ax = fig.add_subplot(121)
ax.plot(history.history["val_loss"])
ax.set_title("validation loss")
ax.set_xlabel("epochs")

ax2 = fig.add_subplot(122)
ax2.plot(history.history["val_acc"])
ax2.set_title("validation accuracy")
ax2.set_xlabel("epochs")
ax2.set_ylim(0, 1)

plt.show()

KeyError: 'val_acc'

Notice that the validation loss begins to actually rise after around 16 epochs, even though validation accuracy remains roughly between 40% and 50%. This suggests our model begins overfitting around then, and best performance would have been achieved if we had stopped early around then. Nevertheless, our accuracy would not have likely been above 50%, and probably lower down.

We can also get a final evaluation by running our model on the training set. Doing so, we get the following results:

In [20]:
loss, accuracy = model.evaluate(x_test, y_test, verbose=0)
print('Test loss:', loss)
print('Test accuracy:', accuracy)

Test loss: 2.640036106109619
Test accuracy: 0.4055793881416321


Finally, we see that we have achieved a (top-1) accuracy of around 49%. That's not too bad for 6000 images, considering that if we were to use a naive strategy of taking random guesses, we would have only gotten around 1% accuracy.

## Transfer learning by starting with existing network

Now we can move on to the main strategy for training an image classifier on our small dataset: by starting with a larger and already trained network.

To start, we will load the VGG16 from keras, which was trained on ImageNet and the weights saved online. If this is your first time loading VGG16, you'll need to wait a bit for the weights to download from the web. Once the network is loaded, we can again inspect the layers with the `summary()` method.

In [21]:
vgg = keras.applications.VGG16(weights='imagenet', include_top=True)
vgg.summary()

553467096/553467096 ━━━━━━━━━━━━━━━━━━━━ 895s 2us/step


Model: "vgg16"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ input_layer_1 (InputLayer)           │ (None, 224, 224, 3)         │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block1_conv1 (Conv2D)                │ (None, 224, 224, 64)        │           1,792 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block1_conv2 (Conv2D)                │ (None, 224, 224, 64)        │          36,928 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block1_pool (MaxPooling2D)           │ (None, 112, 112, 64)        │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block2_conv1 (Conv2D)                │ (None, 112, 112, 128)       │          73,856 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block2_conv2 (Conv2D)                │ (None, 112, 112, 128)       │         147,584 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block2_pool (MaxPooling2D)           │ (None, 56, 56, 128)         │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block3_conv1 (Conv2D)                │ (None, 56, 56, 256)         │         295,168 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block3_conv2 (Conv2D)                │ (None, 56, 56, 256)         │         590,080 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block3_conv3 (Conv2D)                │ (None, 56, 56, 256)         │         590,080 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block3_pool (MaxPooling2D)           │ (None, 28, 28, 256)         │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block4_conv1 (Conv2D)                │ (None, 28, 28, 512)         │       1,180,160 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block4_conv2 (Conv2D)                │ (None, 28, 28, 512)         │       2,359,808 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block4_conv3 (Conv2D)                │ (None, 28, 28, 512)         │       2,359,808 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block4_pool (MaxPooling2D)           │ (None, 14, 14, 512)         │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block5_conv1 (Conv2D)                │ (None, 14, 14, 512)         │       2,359,808 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block5_conv2 (Conv2D)                │ (None, 14, 14, 512)         │       2,359,808 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block5_conv3 (Conv2D)                │ (None, 14, 14, 512)         │       2,359,808 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block5_pool (MaxPooling2D)           │ (None, 7, 7, 512)           │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ flatten (Flatten)                    │ (None, 25088)               │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ fc1 (Dense)                          │ (None, 4096)                │     102,764,544 │
├──────────────────────────────────────┼─────────────────────────────┼──────────────

 Total params: 138,357,544 (527.79 MB)

 Trainable params: 138,357,544 (527.79 MB)

 Non-trainable params: 0 (0.00 B)

Notice that VGG16 is _much_ bigger than the network we constructed earlier. It contains 13 convolutional layers and two fully connected layers at the end, and has over 138 million parameters, around 100 times as many parameters than the network we made above. Like our first network, the majority of the parameters are stored in the connections leading into the first fully-connected layer.

VGG16 was made to solve ImageNet, and achieves a [8.8% top-5 error rate](https://github.com/jcjohnson/cnn-benchmarks), which means that 91.2% of test samples were classified correctly within the top 5 predictions for each image. It's top-1 accuracy--equivalent to the accuracy metric we've been using (that the top prediction is correct)--is 73%. This is especially impressive since there are not just 97, but 1000 classes, meaning that random guesses would get us only 0.1% accuracy.

In order to use this network for our task, we "remove" the final classification layer, the 1000-neuron softmax layer at the end, which corresponds to ImageNet, and instead replace it with a new softmax layer for our dataset, which contains 97 neurons in the case of the 101_ObjectCategories dataset.

In terms of implementation, it's easier to simply create a copy of VGG from its input layer until the second to last layer, and then work with that, rather than modifying the VGG object directly. So technically we never "remove" anything, we just circumvent/ignore it. This can be done in the following way, by using the keras `Model` class to initialize a new model whose input layer is the same as VGG but whose output layer is our new softmax layer, called `new_classification_layer`. Note: although it appears we are duplicating this large network, internally Keras is actually just copying all the layers by reference, and thus we don't need to worry about overloading the memory.

In [22]:
# make a reference to VGG's input layer
inp = vgg.input

# make a new softmax layer with num_classes neurons
new_classification_layer = Dense(num_classes, activation='softmax')

# connect our new layer to the second to last layer in VGG, and make a reference to it
out = new_classification_layer(vgg.layers[-2].output)

# create a new network between inp and out
model_new = Model(inp, out)


We are going to retrain this network, `model_new` on the new dataset and labels. But first, we need to freeze the weights and biases in all the layers in the network, except our new one at the end, with the expectation that the features that were learned in VGG should still be fairly relevant to the new image classification task. Not optimal, but most likely better than what we can train to in our limited dataset.

By setting the `trainable` flag in each layer false (except our new classification layer), we ensure all the weights and biases in those layers remain fixed, and we simply train the weights in the one layer at the end. In some cases, it is desirable to *not* freeze all the pre-classification layers. If your dataset has enough samples, and doesn't resemble ImageNet very much, it might be advantageous to fine-tune some of the VGG layers along with the new classifier, or possibly even all of them. To do this, you can change the below code to make more of the layers trainable.

In the case of CalTech-101, we will just do feature extraction, fearing that fine-tuning too much with this dataset may overfit. But maybe we are wrong? A good exercise would be to try out both, and compare the results.

So we go ahead and freeze the layers, and compile the new model with exactly the same optimizer and loss function as in our first network, for the sake of a fair comparison. We then run `summary` again to look at the network's architecture.

In [23]:
# make all layers untrainable by freezing weights (except for last layer)
for l, layer in enumerate(model_new.layers[:-1]):
    layer.trainable = False

# ensure the last layer is trainable/not frozen
for l, layer in enumerate(model_new.layers[-1:]):
    layer.trainable = True

model_new.compile(loss='categorical_crossentropy',
              optimizer='adam',
              metrics=['accuracy'])

model_new.summary()

Model: "functional_20"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ input_layer_1 (InputLayer)           │ (None, 224, 224, 3)         │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block1_conv1 (Conv2D)                │ (None, 224, 224, 64)        │           1,792 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block1_conv2 (Conv2D)                │ (None, 224, 224, 64)        │          36,928 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block1_pool (MaxPooling2D)           │ (None, 112, 112, 64)        │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block2_conv1 (Conv2D)                │ (None, 112, 112, 128)       │          73,856 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block2_conv2 (Conv2D)                │ (None, 112, 112, 128)       │         147,584 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block2_pool (MaxPooling2D)           │ (None, 56, 56, 128)         │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block3_conv1 (Conv2D)                │ (None, 56, 56, 256)         │         295,168 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block3_conv2 (Conv2D)                │ (None, 56, 56, 256)         │         590,080 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block3_conv3 (Conv2D)                │ (None, 56, 56, 256)         │         590,080 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block3_pool (MaxPooling2D)           │ (None, 28, 28, 256)         │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block4_conv1 (Conv2D)                │ (None, 28, 28, 512)         │       1,180,160 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block4_conv2 (Conv2D)                │ (None, 28, 28, 512)         │       2,359,808 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block4_conv3 (Conv2D)                │ (None, 28, 28, 512)         │       2,359,808 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block4_pool (MaxPooling2D)           │ (None, 14, 14, 512)         │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block5_conv1 (Conv2D)                │ (None, 14, 14, 512)         │       2,359,808 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block5_conv2 (Conv2D)                │ (None, 14, 14, 512)         │       2,359,808 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block5_conv3 (Conv2D)                │ (None, 14, 14, 512)         │       2,359,808 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block5_pool (MaxPooling2D)           │ (None, 7, 7, 512)           │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ flatten (Flatten)                    │ (None, 25088)               │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ fc1 (Dense)                          │ (None, 4096)                │     102,764,544 │
├──────────────────────────────────────┼─────────────────────────────┼──────────────

 Total params: 134,657,953 (513.68 MB)

 Trainable params: 397,409 (1.52 MB)

 Non-trainable params: 134,260,544 (512.16 MB)

Looking at the summary, we see the network is identical to the VGG model we instantiated earlier, except the last layer, formerly a 1000-neuron softmax, has been replaced by a new 97-neuron softmax. Additionally, we still have roughly 134 million weights, but now the vast majority of them are "non-trainable params" because we froze the layers they are contained in. We now only have 397,000 trainable parameters, which is actually only a quarter of the number of parameters needed to train the first model.

As before, we go ahead and train the new model, using the same hyperparameters (batch size and number of epochs) as before, along with the same optimization algorithm. We also keep track of its history as we go.

In [24]:
history2 = model_new.fit(x_train, y_train,
                         batch_size=128,
                         epochs=10,
                         validation_data=(x_val, y_val))


Epoch 1/10
34/34 ━━━━━━━━━━━━━━━━━━━━ 1200s 35s/step - accuracy: 0.1622 - loss: 4.0949 - val_accuracy: 0.3287 - val_loss: 3.0410
Epoch 2/10
34/34 ━━━━━━━━━━━━━━━━━━━━ 1196s 35s/step - accuracy: 0.4510 - loss: 2.5279 - val_accuracy: 0.5295 - val_loss: 2.2128
Epoch 3/10
34/34 ━━━━━━━━━━━━━━━━━━━━ 1169s 35s/step - accuracy: 0.5748 - loss: 1.9271 - val_accuracy: 0.5725 - val_loss: 1.8619
Epoch 4/10
34/34 ━━━━━━━━━━━━━━━━━━━━ 1172s 35s/step - accuracy: 0.6364 - loss: 1.6202 - val_accuracy: 0.5843 - val_loss: 1.6923
Epoch 5/10
28/34 ━━━━━━━━━━━━━━━━━━━━ 2:50 28s/step - accuracy: 0.6637 - loss: 1.4625

KeyboardInterrupt: 

Our validation accuracy hovers close to 80% towards the end, which is more than 30% improvement on the original network trained from scratch (meaning that we make the wrong prediction on 20% of samples, rather than 50%).

It's worth noting also that this network actually trains _slightly faster_ than the original network, despite having more than 100 times as many parameters! This is because freezing the weights negates the need to backpropagate through all those layers, saving us on runtime.

Let's plot the validation loss and accuracy again, this time comparing the original model trained from scratch (in blue) and the new transfer-learned model in green.

In [ ]:
fig = plt.figure(figsize=(16,4))
ax = fig.add_subplot(121)
ax.plot(history.history["val_loss"])
ax.plot(history2.history["val_loss"])
ax.set_title("validation loss")
ax.set_xlabel("epochs")

ax2 = fig.add_subplot(122)
ax2.plot(history.history["val_acc"])
ax2.plot(history2.history["val_acc"])
ax2.set_title("validation accuracy")
ax2.set_xlabel("epochs")
ax2.set_ylim(0, 1)

plt.show()

In [6]:
!pip -q install gdown

# ID do arquivo no seu link: 1bCJaQE-rVER2JwbaYwwjzNO_AuXDikbQ
!echo "Downloading 101_Object_Categories for image notebooks"
!gdown --id 1bCJaQE-rVER2JwbaYwwjzNO_AuXDikbQ -O 101_ObjectCategories.tar.gz
!tar -xzf 101_ObjectCategories.tar.gz
!rm 101_ObjectCategories.tar.gz
!ls



[notice] A new release of pip is available: 24.3.1 -> 25.2
[notice] To update, run: python.exe -m pip install --upgrade pip


"Downloading 101_Object_Categories for image notebooks"


C:\Users\poger\.conda\envs\AI\Lib\site-packages\gdown\__main__.py:140: FutureWarning: Option `--id` was deprecated in version 4.3.1 and will be removed in 5.0. You don't need to pass it anymore to use a file ID.
  warnings.warn(
Downloading...
From (original): https://drive.google.com/uc?id=1bCJaQE-rVER2JwbaYwwjzNO_AuXDikbQ
From (redirected): https://drive.google.com/uc?id=1bCJaQE-rVER2JwbaYwwjzNO_AuXDikbQ&confirm=t&uuid=b3e1e03a-523b-49dd-b8f4-e1caa723bedc
To: C:\Users\poger\101_ObjectCategories.tar.gz

  0%|          | 0.00/132M [00:00<?, ?B/s]
  0%|          | 524k/132M [00:01<05:31, 396kB/s]
  1%|          | 1.05M/132M [00:02<05:47, 376kB/s]
  1%|1         | 1.57M/132M [00:03<05:10, 419kB/s]
  2%|1         | 2.10M/132M [00:04<04:31, 477kB/s]
  2%|1         | 2.62M/132M [00:05<04:24, 488kB/s]
  2%|2         | 3.15M/132M [00:07<05:18, 404kB/s]
  3%|2         | 3.67M/132M [00:07<04:09, 514kB/s]
  3%|3         | 4.19M/132M [00:10<06:41, 317kB/s]
  4%|3         | 4.72M/132M [00:14<09:18

-1.14-windows.xml
101_ObjectCategories
1_lab1.ipynb
3D Objects
Ambiente de Impressão
Ambiente de Rede
AppData
Application Data
CascadeProjects
Configurações Locais
Contacts
Cookies
Dados de Aplicativos
Desktop
Documents
Downloads
Exercicios
Favorites
InstallAnywhere
IntelGraphicsProfiles
Links
Menu Iniciar
Meus Documentos
Modelos
Music
NTUSER.DAT
NTUSER.DAT{1a566ac0-8f5d-11f0-8a47-005056c00008}.TM.blf
NTUSER.DAT{1a566ac0-8f5d-11f0-8a47-005056c00008}.TMContainer00000000000000000001.regtrans-ms
NTUSER.DAT{1a566ac0-8f5d-11f0-8a47-005056c00008}.TMContainer00000000000000000002.regtrans-ms
OneDrive
Oracle
Pictures
Postman
Postman Agent
Recent
Saved Games
Searches
SendTo
Sti_Trace.log
Untitled.ipynb
Videos
ai_completion
anaconda_projects
ansel
cachedir_joblib
chroma.log
delete
demo_find_face.py
demo_find_green_blob.py
dwhelper
helloworld_1.py
ipython_hist.sqlite
json
maxima
migrations_202504181229.txt
my_chroma_data
ntuser.dat.LOG1
ntuser.dat.LOG2
ntuser.ini
requirements.txt
scoop
source
s

In [6]:
!pip -q install gdown

# ID do arquivo no seu link: 1bCJaQE-rVER2JwbaYwwjzNO_AuXDikbQ
!echo "Downloading 101_Object_Categories for image notebooks"
!gdown --id 1bCJaQE-rVER2JwbaYwwjzNO_AuXDikbQ -O 101_ObjectCategories.tar.gz
!tar -xzf 101_ObjectCategories.tar.gz
!rm 101_ObjectCategories.tar.gz
!ls



[notice] A new release of pip is available: 24.3.1 -> 25.2
[notice] To update, run: python.exe -m pip install --upgrade pip


"Downloading 101_Object_Categories for image notebooks"


C:\Users\poger\.conda\envs\AI\Lib\site-packages\gdown\__main__.py:140: FutureWarning: Option `--id` was deprecated in version 4.3.1 and will be removed in 5.0. You don't need to pass it anymore to use a file ID.
  warnings.warn(
Downloading...
From (original): https://drive.google.com/uc?id=1bCJaQE-rVER2JwbaYwwjzNO_AuXDikbQ
From (redirected): https://drive.google.com/uc?id=1bCJaQE-rVER2JwbaYwwjzNO_AuXDikbQ&confirm=t&uuid=b3e1e03a-523b-49dd-b8f4-e1caa723bedc
To: C:\Users\poger\101_ObjectCategories.tar.gz

  0%|          | 0.00/132M [00:00<?, ?B/s]
  0%|          | 524k/132M [00:01<05:31, 396kB/s]
  1%|          | 1.05M/132M [00:02<05:47, 376kB/s]
  1%|1         | 1.57M/132M [00:03<05:10, 419kB/s]
  2%|1         | 2.10M/132M [00:04<04:31, 477kB/s]
  2%|1         | 2.62M/132M [00:05<04:24, 488kB/s]
  2%|2         | 3.15M/132M [00:07<05:18, 404kB/s]
  3%|2         | 3.67M/132M [00:07<04:09, 514kB/s]
  3%|3         | 4.19M/132M [00:10<06:41, 317kB/s]
  4%|3         | 4.72M/132M [00:14<09:18

-1.14-windows.xml
101_ObjectCategories
1_lab1.ipynb
3D Objects
Ambiente de Impressão
Ambiente de Rede
AppData
Application Data
CascadeProjects
Configurações Locais
Contacts
Cookies
Dados de Aplicativos
Desktop
Documents
Downloads
Exercicios
Favorites
InstallAnywhere
IntelGraphicsProfiles
Links
Menu Iniciar
Meus Documentos
Modelos
Music
NTUSER.DAT
NTUSER.DAT{1a566ac0-8f5d-11f0-8a47-005056c00008}.TM.blf
NTUSER.DAT{1a566ac0-8f5d-11f0-8a47-005056c00008}.TMContainer00000000000000000001.regtrans-ms
NTUSER.DAT{1a566ac0-8f5d-11f0-8a47-005056c00008}.TMContainer00000000000000000002.regtrans-ms
OneDrive
Oracle
Pictures
Postman
Postman Agent
Recent
Saved Games
Searches
SendTo
Sti_Trace.log
Untitled.ipynb
Videos
ai_completion
anaconda_projects
ansel
cachedir_joblib
chroma.log
delete
demo_find_face.py
demo_find_green_blob.py
dwhelper
helloworld_1.py
ipython_hist.sqlite
json
maxima
migrations_202504181229.txt
my_chroma_data
ntuser.dat.LOG1
ntuser.dat.LOG2
ntuser.ini
requirements.txt
scoop
source
s

Notice that whereas the original model began overfitting around epoch 16, the new model continued to slowly decrease its loss over time, and likely would have improved its accuracy slightly with more iterations. The new model made it to roughly 80% top-1 accuracy (in the validation set) and continued to improve slowly through 100 epochs.

It's possibly we could have improved the original model with better regularization or more dropout, but we surely would not have made up the >30% improvement in accuracy.

Again, we do a final validation on the test set.

In [25]:
loss, accuracy = model_new.evaluate(x_test, y_test, verbose=0)

print('Test loss:', loss)
print('Test accuracy:', accuracy)

Test loss: 1.5386172533035278
Test accuracy: 0.6437768340110779


To predict a new image, simply run the following code to get the probabilities for each class.

In [28]:
img, x = get_image('101_ObjectCategories/airplanes/image_0003.jpg')
probabilities = model_new.predict([x])


1/1 ━━━━━━━━━━━━━━━━━━━━ 1s 1s/step   


### Improving the results

78.2% top-1 accuracy on 97 classes, roughly evenly distributed, is a pretty good achievement. It is not quite as impressive as the original VGG16 which achieved 73% top-1 accuracy on 1000 classes. Nevertheless, it is much better than what we were able to achieve with our original network, and there is room for improvement. Some techniques which possibly could have improved our performance.

- Using data augementation: augmentation refers to using various modifications of the original training data, in the form of distortions, rotations, rescalings, lighting changes, etc to increase the size of the training set and create more tolerance for such distortions.
- Using a different optimizer, adding more regularization/dropout, and other hyperparameters.
- Training for longer (of course)

A more advanced example of transfer learning in Keras, involving augmentation for a small 2-class dataset, can be found in the [Keras blog](https://blog.keras.io/building-powerful-image-classification-models-using-very-little-data.html).